builtin/providers/aws/resource_aws_s3_bucket_object.go

```diff
@@ -74,6 +75,11 @@ func resourceAwsS3BucketObject() *schema.Resource {
 				ConflictsWith: []string{"source"},
 			},
 
+			"kms_key_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
 			"etag": &schema.Schema{
 				Type: schema.TypeString,
 				// This will conflict with SSE-C and SSE-KMS encryption and multi-part upload
```

Comment:
Since we cannot use `ValidateFunc` to guard against this (until https://github.com/hashicorp/terraform/pull/4348 is finished/merged), I think we'll have to add some conditionals into `Create` and `Update`.

Essentially we want to prevent the user from specifying `etag` when they specified `kms_key_id`, because there's no way they can calculate that (unless they know how AWS encrypts the data via KMS).

It would be great if we could also somehow prevent them from using the calculated value, as that's useless too, but I'm not quite sure how to approach that.